### PR TITLE
Introducing densitymapbox traces

### PR DESCRIPTION
**Pull request by etpinard**
Based https://github.com/plotly/plotly.js/pull/3988 (which is based off https://github.com/plotly/plotly.js/pull/3987)

Resolves https://github.com/plotly/plotly.js/issues/3110

---------

This PR adds a new mapbox-based trace type: `densitymapbox` 

Example: https://codepen.io/etpinard/pen/KjZXOr

`densitymapbox` traces are essentially just a wrapper around [`mapbox-gl`'s "heatmap" layers](https://docs.mapbox.com/mapbox-gl-js/example/heatmap-layer/) - which we chose not the call _`heatmapmapbox`_ as their algorithm is "too different" from our `heatmap` traces.

Please refer to the comments and descriptions in https://github.com/plotly/plotly.js/commit/7e15d817deff6df4f1750728d45ef3533b6e8118 for more info.

---

cc @plotly/plotly_js @nicolaskruchten @chriddyp

### CODE REVIEW

File: src/traces/densitymapbox/plot.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+        subplot.setOptions(id, 'setLayoutProperty', opts.layout);
+
+        if(opts.layout.visibility === 'visible') {
+            subplot.setOptions(id, 'setPaintProperty', opts.paint);
+        }
+    }
+};
+
+proto._addLayers = function(optsAll) {
```
Comment (antoinerg) [inline src/traces/densitymapbox/plot.js]: Can we reuse the one from `src/traces/choroplethmapbox/plot.js`? I think it is identical

File: src/traces/densitymapbox/plot.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    ];
+
+    // previous 'below' value,
+    // need this to update it properly
+    this.below = null;
+}
+
+var proto = DensityMapbox.prototype;
+
+proto.update = function(calcTrace) {
```
Comment (antoinerg) [inline src/traces/densitymapbox/plot.js]: Also very similar to the function in `src/traces/choroplethmapbox/plot.js`. I'll leave it up to you to decide whether to reuse it or not.

File: src/traces/densitymapbox/plot.js (hunk 1 above)
Comment (etpinard) [reply src/traces/densitymapbox/plot.js]: We could, but I don't think it's worth it.

File: src/traces/densitymapbox/plot.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    this.below = below;
+};
+
+proto._removeLayers = function() {
+    var map = this.subplot.map;
+    var layerList = this.layerList;
+
+    for(var i = 0; i < layerList.length; i++) {
```
Comment (archmoj) [inline src/traces/densitymapbox/plot.js]: The loop order may be reversed here.
Removing from last item of the list towards the first item may have the benefit of having the valid index. 
On another note - I was wondering if the list should be set to an empty array at the end of this process?

Comment (etpinard) [reply src/traces/densitymapbox/plot.js]: Done in https://github.com/plotly/plotly.js/pull/3993/commits/1add90efdfb7d930ff3c0c880199cdeef776b5e2

**Comment by antoinerg** [comment]
It looks great :tada: ! Nice job @etpinard :muscle: 

 :dancer:

**Comment by meghnakayam** [comment]
This looks great. Is there a way to incorporate the same in python for Plotly Dash?

**Comment by nicolaskruchten** [comment]
@meghnakayam yep! This is coming to Plotly.py/Dash in a couple of weeks :)

**Pull request merged**